File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def solve(self , i , v , c , t , ans ):
        if i >= len(c):
            if(t == 0):
                ans.append(v[:])
            return 
        
        if c[i] <= t:
            v.append(c[i])
            self.solve(i+1 , v , c , t - c[i] , ans)
            v.pop()
        j = i + 1
        while j < len(c) and c[j] == c[i]:
            j += 1
        self.solve(j, v, c, t, ans)


    def combinationSum2(self, c: List[int], t: int) -> List[List[int]]:
        ans = [] 
        v = []
        c.sort()
        self.solve(0 , v , c , t , ans)
        return ans

File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
    void solve(int i,vector<int>&v , int t, vector<int>& c,vector<vector<int>> &ans ){
        if(i >= c.size()){
            if(t == 0)
                ans.push_back(v);
            return ;
        }
        if(c[i] <= t){
            v.push_back(c[i]);
            solve(i+1 , v , t - c[i] , c ,ans);
            v.pop_back();
        }
        while(i+1 < c.size() && c[i] == c[i+1])
            i++;
        solve(i+1 , v , t , c ,ans);
    }
    vector<vector<int>> combinationSum2(vector<int>& c, int t) {
        vector<vector<int>> ans;
        vector<int> v;
        sort(c.begin(),c.end());

        solve(0 , v , t , c , ans);
        return ans;

    }
};